### DISCUSSION TITLE
How to build bazel release?

### DISCUSSION THREAD

**Discussion by bhuntsman** [bazelbuild/bazel#22882] [General]
I'm trying to build jax.  The Jax script uses bazel --version to check if the bazel version is sufficient.  However, if I build bazel from git, --version returns "no_version".  I've tried this on a few different platforms and releases.

How exactly does someone build bazel in such a way that --version works?  This can't be that obscure of an issue, but I can't find any documentation to indicate what should be done.

Thank you!

**Comment by meteorcloudy** [comment]
You'll need [`--embed_label`](https://bazel.build/reference/command-line-reference#flag--embed_label)=7.2.0. I think JAX can also treat no_version as HEAD and don't fail the check?

**Reply by bhuntsman** [reply]
Hi @meteorcloudy, thanks for the reply!
Is that the correct way to make a production build from the bazel git repo?  It looks to me like there's some amount of logic in compile.py that should extract some info from git based on the HEAD and use that.  I couldn't find any documentation for how to build a production bazel executable, just bazel-dev, and that doesn't work right with --version either.

It's not explicitly a jax issue.  But no, jax most certainly does not have logic to treat no_version as head.  If it doesn't get an actual version string it fails the build out almost immediately.

Thanks again!!

**Reply by meteorcloudy** [reply]
> Is that the correct way to make a production build from the bazel git repo?

Yes, we do use this flag in our [release build](https://github.com/bazelbuild/continuous-integration/blob/ee5ea988681e086afabfe6677eef2dc1292f6b9d/pipelines/bazel-release.yml#L91) and others like --stamp and --workspace_status_command=scripts/ci/build_status_command.sh. 

>  If it doesn't get an actual version string it fails the build out almost immediately.

I think it shouldn't fail the build but print out a warning like TensorFlow does: https://github.com/tensorflow/tensorflow/blob/1b5fc0570edac4d7b82435b3612e402f9594cf14/configure.py#L444

**Reply by bhuntsman** [reply]
So for people wanting to create a release build from git, the correct thing is to bootstrap, then:

```
bazel build //src:bazel --compilation_mode=opt --embed_label=7.2.0
```

correct?  Or is there more of a procedure?

As for the jax issue, it's unrelated to the issue at hand, but yes you are correct, it doesn't fail, it prints the following and then quits:
```
Cannot find or download a suitable version of bazel."
        "Please install bazel >= 6.5.0.
```

Thank you!

**Reply by meteorcloudy** [reply]
Yes, that's correct.

**Reply by bhuntsman** [reply]
Hi @meteorcloudy I just did a build using that command above, and --version still returns no_version.  Any other ideas?

Thank you.

**Reply by meteorcloudy** [reply]
You probably also need `--stamp --workspace_status_command=scripts/ci/build_status_command.sh`.

**Reply by meteorcloudy** [reply]
I'll update https://bazel.build/install/compile-source to include this instruction.

**Reply by bhuntsman** [reply]
Thank you @meteorcloudy, that got me past that issue and I can now build a bazel binary from the git repo that has a version string in it.  However, the next problem is that despite compiling it with JDK 21 per the instructions, it insists on downloading and trying to run the results under Java 20, and gets the following error:

```
...
Starting local Bazel server and connecting to it...
Server crashed during startup. Now printing /home/build/.cache/bazel/_bazel_build/dcc0c88d87f0eff69c7a10c9d397572b/server/jvm.out
OpenJDK 64-Bit Server VM warning: Options -Xverify:none and -noverify were deprecated in JDK 13 and will likely be removed in a future release.
Error: LinkageError occurred while loading main class com.google.devtools.build.lib.bazel.Bazel
	java.lang.UnsupportedClassVersionError: com/google/devtools/build/lib/bazel/Bazel has been compiled by a more recent version of the Java Runtime (class file version 65.0), this version of the Java Runtime only recognizes class file versions up to 64.0
...
```

Running the java that it procured somehow shows that it's the wrong version:
```
$ /home/build/.cache/bazel/_bazel_build/install/0affd1f510016b459c381cdc6385fa98/embedded_tools/jdk/bin/java --version
openjdk 20 2023-03-21
OpenJDK Runtime Environment Temurin-20+36 (build 20+36)
OpenJDK 64-Bit Server VM Temurin-20+36 (build 20+36, mixed mode)
```

Is there a way to force it to either use the locally-installed jdk 21 by default (so I don't have to add additional command line arguments), or make the build embed (or fetch) the correct version of Java (21, not 20)?

Thank you so much!

**Reply by bhuntsman** [reply]
Looks like this could possibly be solved by mirroring the following URL to mirror.bazel.build:
https://github.com/adoptium/temurin21-binaries/releases/download/jdk-21.0.3%2B9/OpenJDK21U-jdk_ppc64le_linux_hotspot_21.0.3_9.tar.gz

and also the following diff:
```
diff --git a/repositories.bzl b/repositories.bzl
index d73fed5653..047de29da1 100644
--- a/repositories.bzl
+++ b/repositories.bzl
@@ -92,12 +92,11 @@ def embedded_jdk_repositories():
         url = "https://github.com/adoptium/temurin19-binaries/releases/download/jdk-19.0.2%2B7/OpenJDK19U-jdk_s390x_linux_hotspot_19.0.2_7.tar.gz",
     )
 
-    # JDK21 unavailable so use JDK19 instead for linux ppc64le.
     http_file(
         name = "openjdk_linux_ppc64le_vanilla",
-        sha256 = "45dde71faf8cbb78fab3c976894259655c8d3de827347f23e0ebe5710921dded",
+        sha256 = "9a1079d7f0fc72951fdc9a0029e49a15f6ba114683aee626f882ee2c761f1d57",
         downloaded_file_path = "adoptopenjdk-ppc64le-vanilla.tar.gz",
-        url = "https://github.com/adoptium/temurin20-binaries/releases/download/jdk-20%2B36/OpenJDK20U-jdk_ppc64le_linux_hotspot_20_36.tar.gz",
+        url = "https://github.com/adoptium/temurin21-binaries/releases/download/jdk-21.0.3%2B9/OpenJDK21U-jdk_ppc64le_linux_hotspot_21.0.3_9.tar.gz",
     )
     http_file(
         name = "openjdk_macos_x86_64_vanilla",
```

Any chance you could at least do the mirroring, @meteorcloudy or @meisterT?

**Reply by sluongng** [reply]
@bhuntsman there is an issue template for mirroring requests here https://github.com/bazelbuild/bazel/issues/new?assignees=sgowroji%2Ciancha1992%2Csatyanandak&labels=mirror+request%2Ctype%3A+process%2CP2%2Cteam-OSS&projects=&template=mirror_request.yml&title=%5BMirror%5D+

**Reply by bhuntsman** [reply]
Thank you @sluongng, I submitted a mirroring request.

**Reply by meteorcloudy** [reply]
Please also feel free to send the diff as the PR.

**Comment by meisterT** [comment]
@bhuntsman I don't know what the context of your question is. Are you aware of our rolling releases? https://bazel.build/release/rolling

**Reply by bhuntsman** [reply]
Hi @meisterT!  Yes, I am aware of those.  However, you don't currently ship binaries for my platform (Linux ppc64le), and the app I'm trying to compile checks the bazel --version output.  So I need a way to generate a bazel that returns a numeric version string and not no_version.  I also need to be able to build from the git checkout and not the dist zip in case I need to make any source changes.  It would be nice if one could just check out the git repo, compile it, and end up with a bazel that didn't say "no_version".  I think maybe just the documentation isn't clear on how to build it in a "release" fashion.  But @meteorcloudy's suggestion of --embed_label should get me far enough down the road that I can concentrate on what I'm trying compile.

Thank you much!
